### PR TITLE
Features/cassandra 3.x plugin

### PR DESCRIPTION
**Pull request by stone-wlg**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

### New feature or improvement
http://cassandra.apache.org/

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please solve the conflict.

**Comment by wu-sheng** [comment]
@michaelsembwever I think you are the best option for us to review this plugin :) SkyWalking plugin for Cassandra Java driver. We will require integration tests, but basically, I hope you could help to review that, we don't miss anything.

File: apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        <groupId>org.apache.skywalking</groupId>
+        <version>6.4.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>apm-cassandra-3.x-plugin</artifactId>
+    <name>apm-cassandra-3.x-plugin</name>
+
+    <properties>
+        <compiler.version>1.8</compiler.version>
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/pom.xml]: If your plugin requires 1.8+ compiler, this plugin must be in optional-plugin. If your codes don't require 1.8, although the target driver is 1.8+, you don't need to set this.

Comment (stone-wlg) [reply apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/pom.xml]: ok

File: apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.cassandra.v3;
+
+import jnr.ffi.annotations.In;
```
Comment (zhaoyuguang) [inline apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java]: may unused import statement

Comment (stone-wlg) [reply apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java]: ok

File: apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class ClusterConstructorWithStateListenerArgInterceptor implements InstanceConstructorInterceptor {
+    @Override
+    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        List<InetSocketAddress> inetSocketAddresses = (List<InetSocketAddress>) allArguments[1];
+        List<String> hosts = new ArrayList<String>();
+        for (InetSocketAddress inetSocketAddress : inetSocketAddresses) {
+            hosts.add(inetSocketAddress.getHostName() + ":" + inetSocketAddress.getPort());
+        }
+        String contactPoints = StringUtil.join(',', hosts.toArray(new String[0]));
```
Comment (zhaoyuguang) [inline apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java]: maybe use `StringBuilder` be better, like:
```
public class JedisClusterConstructorWithListHostAndPortArgInterceptor implements InstanceConstructorInterceptor {

    @Override
    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
        StringBuilder redisConnInfo = new StringBuilder();
        Set<HostAndPort> hostAndPorts = (Set<HostAndPort>)allArguments[0];
        for (HostAndPort hostAndPort : hostAndPorts) {
            redisConnInfo.append(hostAndPort.toString()).append(";");
        }

        objInst.setSkyWalkingDynamicField(PeerFormat.shorten(redisConnInfo.toString()));
    }
}
```

**Comment by wu-sheng** [comment]
@stone-wlg please fix CI and submodule confilct.

File: apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java (hunk 3 above)
Comment (stone-wlg) [reply apm-sniffer/apm-sdk-plugin/cassandra-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/cassandra/v3/ClusterConstructorWithStateListenerArgInterceptor.java]: ok

**Comment by stone-wlg** [comment]
> @stone-wlg please fix CI and submodule confilct.

Hi, i am doing CI, but about skywalking-ui conflicts i could not see the detail. skywalking-ui is a submodule, should i clone it independently and make commits ? or just follow the skywalking https://github.com/apache/skywalking/blob/v6.3.0/docs/en/guides/How-to-build.md, then make commits on it, but i did it and no conflicts when i merge features/cassandra-3.x-plugin to master on  my fork skywalking.

**Comment by wu-sheng** [comment]
@stone-wlg skywalking-ui is a git submodule, you need make sure it makes commitid. Compile and git conflict are two different things.

**Comment by zhaoyuguang** [comment]
Hi @stone-wlg, Please add testcases at https://github.com/SkyAPMTest/agent-auto-integration-testcases

File: .gitmodules [hunk 4]
```diff
@@ -6,4 +6,4 @@
 	url = https://github.com/apache/skywalking-query-protocol.git
 [submodule "skywalking-ui"]
 	path = skywalking-ui
-	url = https://github.com/apache/skywalking-rocketbot-ui.git
+	url = https://github.com/stone-wlg/skywalking-rocketbot-ui.git
```
Comment (wu-sheng) [inline .gitmodules]: Why do you change this?

**Comment by stone-wlg** [comment]
> @stone-wlg skywalking-ui is a git submodule, you need make sure it makes commitid. Compile and git conflict are two different things.

Hi @wu-sheng , i am confusing that how to submit skywalking-ui changes, because i have no rights to submit the changes on skywalking-rocketbot-ui project, so could i make a fork on skywalking-rocketbot-ui project, and commit changes on my forked skywalking-rocketbot-ui, then ask pull request from my forked skywalking-rocketbot-ui to official skywalking-rocketbot-ui first ?  or is there any instruction that i can follow ?

File: .gitmodules (hunk 4 above)
Comment (stone-wlg) [reply .gitmodules]: > Why do you change this?

rollback, just for test.

**Comment by wu-sheng** [comment]
You don't need any change to ui module, from my understanding. If you do, that change should be done by another PR to there directly.

Current error is, you are changing submodule, which is unexpected. You just need to make sure your submodule commit id equals master branch.

**Comment by stone-wlg** [comment]
> You don't need any change to ui module, from my understanding. If you do, that change should be done by another PR to there directly.
> 
> Current error is, you are changing submodule, which is unexpected. You just need to make sure your submodule commit id equals master branch.

ok, i do not need to change UI project even i add new plugin, but there will be some elements missing on UI, is there somebody to fix it?

**Comment by kezhenxu94** [comment]
moved to #3330

**Pull request closed without merging**